<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Models-Discussion" data-toc-modified-id="Models-Discussion-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Models Discussion</a></span><ul class="toc-item"><li><span><a href="#¿Why-we-don't-implement-linear-regression?" data-toc-modified-id="¿Why-we-don't-implement-linear-regression?-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>¿Why we don't implement linear regression?</a></span></li><li><span><a href="#Performance-Metrics" data-toc-modified-id="Performance-Metrics-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Performance Metrics</a></span><ul class="toc-item"><li><span><a href="#Performance-metric-for-banknote-authentication-data-set" data-toc-modified-id="Performance-metric-for-banknote-authentication-data-set-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Performance metric for banknote authentication data set</a></span></li><li><span><a href="#Performance-metric-for-occupancy-detection-data-set" data-toc-modified-id="Performance-metric-for-occupancy-detection-data-set-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Performance metric for occupancy detection data set</a></span></li></ul></li></ul></li><li><span><a href="#Data-Processing" data-toc-modified-id="Data-Processing-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Processing</a></span><ul class="toc-item"><li><span><a href="#Banknote-authentication-data-set-processing" data-toc-modified-id="Banknote-authentication-data-set-processing-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Banknote authentication data set processing</a></span></li><li><span><a href="#Occupancy-detection-data-set-procesing" data-toc-modified-id="Occupancy-detection-data-set-procesing-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Occupancy detection data set procesing</a></span></li></ul></li><li><span><a href="#Models-Implementation" data-toc-modified-id="Models-Implementation-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Models Implementation</a></span><ul class="toc-item"><li><span><a href="#Logistic-Regression" data-toc-modified-id="Logistic-Regression-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Logistic Regression</a></span><ul class="toc-item"><li><span><a href="#Performance-metrics-by-confusion-matrix-example" data-toc-modified-id="Performance-metrics-by-confusion-matrix-example-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Performance metrics by confusion matrix example</a></span></li><li><span><a href="#Performance-metrics-in-banknote-authentication" data-toc-modified-id="Performance-metrics-in-banknote-authentication-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Performance metrics in banknote authentication</a></span></li><li><span><a href="#Performance-metrics-in-occupancy-problem" data-toc-modified-id="Performance-metrics-in-occupancy-problem-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>Performance metrics in occupancy problem</a></span></li></ul></li><li><span><a href="#Decision-tree-learning" data-toc-modified-id="Decision-tree-learning-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Decision tree learning</a></span><ul class="toc-item"><li><span><a href="#Performance-metrics-in-banknote-authentication" data-toc-modified-id="Performance-metrics-in-banknote-authentication-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Performance metrics in banknote authentication</a></span></li><li><span><a href="#Performance-metrics-in-occupancy-detection" data-toc-modified-id="Performance-metrics-in-occupancy-detection-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Performance metrics in occupancy detection</a></span></li></ul></li><li><span><a href="#Support-vector-machine-(SVM)" data-toc-modified-id="Support-vector-machine-(SVM)-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Support vector machine (SVM)</a></span><ul class="toc-item"><li><span><a href="#Grid-Search-for-SVM" data-toc-modified-id="Grid-Search-for-SVM-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Grid Search for SVM</a></span></li><li><span><a href="#Performance-metrics-in-banknote-authentication" data-toc-modified-id="Performance-metrics-in-banknote-authentication-4.3.2"><span class="toc-item-num">4.3.2&nbsp;&nbsp;</span>Performance metrics in banknote authentication</a></span></li><li><span><a href="#Performance-metrics-in-occupancy-detection" data-toc-modified-id="Performance-metrics-in-occupancy-detection-4.3.3"><span class="toc-item-num">4.3.3&nbsp;&nbsp;</span>Performance metrics in occupancy detection</a></span></li></ul></li><li><span><a href="#K---Nearest-neighbors-(KNN)" data-toc-modified-id="K---Nearest-neighbors-(KNN)-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>K - Nearest neighbors (KNN)</a></span><ul class="toc-item"><li><span><a href="#Performance-metrics-in-banknote-authentication" data-toc-modified-id="Performance-metrics-in-banknote-authentication-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Performance metrics in banknote authentication</a></span></li><li><span><a href="#Performance-metrics-in-occupancy-detection" data-toc-modified-id="Performance-metrics-in-occupancy-detection-4.4.2"><span class="toc-item-num">4.4.2&nbsp;&nbsp;</span>Performance metrics in occupancy detection</a></span></li></ul></li></ul></li><li><span><a href="#Results" data-toc-modified-id="Results-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Results</a></span></li><li><span><a href="#Conclusions,-Discussions-and-Future-Work" data-toc-modified-id="Conclusions,-Discussions-and-Future-Work-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Conclusions, Discussions and Future Work</a></span></li></ul></div>

# Introduction

The objective of this project is to explore and learn about machine learning algorithms. The main purpose is to implement a few algorithms (probably the most important ML algorithms) and study their performance in two fixed datasets.

Remember; for this project we are usen the following datasets: <br>
https://archive.ics.uci.edu/ml/datasets/banknote+authentication <br>
https://archive.ics.uci.edu/ml/datasets/Occupancy+Detection+

For a long discussion of these datasets please refer to First_HW jupyter notebook in the main page of the repository. 

# Models Discussion

For each dataset we're going to implement the following models:
<ol type="1">
   <li>Logistic Regression</li>
   <li>Decision Tree Learning</li> 
   <li>Soft Margin Support Vector Machine (Soft-SVM)</li>
   <li>K-Nearest Neighboors (KNN)</li>
</ol>

## ¿Why we don't implement linear regression?

The propose of our learning problem is predict the target outputs for new data inputs. In our problem we have binary (categorical) labels and we dont want adjust our model to the data, instead, we want to split the representation space to distinguish predicted labels of new data points. The labor of a linear regression is to adjust data, no split it. This method would be valuable if blank data appears: we would be able to fill these blank spaces with a linear regression outputs.

## Performance Metrics

Depending of the research or bussiness objective we need to stablish a performance metric that allow us, without any type of bias, compare and decide when a model have a good or bad behavior (in the sense of generalize new data). For this reason, we set the performance metric before we implement any type of model.  This task demands thinking in the problem, and define some losses/gain criteria. <br>

In order to reach a desired metric for each of our two problems, we're gonna analyce them thinking two critical events: false negative and false positive cases. For this two we're gonna decide which have more negative/possite impact in our objective depending on the problem. Remember the false negative, false positive term comes form the confusion matrix concept; the following is how sklearn implement the confusion matrix:

| Actual \ Predicted |  0 |  1 |
|--------------------|:--:|:--:|
|                  0 | TN | FP |
|                  1 | FN | TP |

Even if we define a performance metric for our problem, for purporses of this study, we're gonna also calculate the **accuracy score (AS)** as a standar model performance metric, remember AS is defined as:
$$ Accuracy = \frac{TN + TP}{TN + FN + FP + TP}$$

### Performance metric for banknote authentication data set

In this problem we have the task of detect whenever a banknote is genuine or not. Thinking as a seller; we note the case when a no genuine banknote entry in our cash flow (false positive case) is more important than when a genuine banknote is rejected from our cash flow (false negative case). The first creates negative movement in our cash flow, whereas, second doesn't even create a movement possite or negative movement. For this reasons we can think the false positive case must have bigger weight than a false positive. For this reasons we can conclude that the **precision score (PS)** is a good measure for assess the performance of our models: <br>
$$\text{Precision} = \frac{TP}{FP + TP}$$

Here, we want precision close to one, this means $FP$ must be close to zero, exactly what we want in our problem. <br>
In the data processing phase we'll see that the classes are balanced, and no imbalanced metrics need to be implemented for this problem.

### Performance metric for occupancy detection data set

For this problem, our task is detect if there's someone inside a room or not. This problem can be raised from a security company, they can be interested in detect room occupancy in order to alarm when there's someone in a restricted hour, for example. In this order of ideas, if my security system is non-capable of tell me when there's a person big troubles can araise (this is a false negative case). In the other hand, if there's no one inside but my system alerts of occupancy, we can just turn off alarms and everything will be ok (this is a false positive case). For this reason we can chose **recall score (RS)** as our performance metric for this problem: <br>
$$\text{Recall} = \frac{TP}{FN + TP} $$

Again, we want recall close to one, this means $FN$ must be close to zero. <br>
In the data processing phase we'll see that the classes have mild imbalanced, but we're gonna keep forward with the chosen performance metric to see if we can generalize. 

# Data Processing

For this part we need to analyze the data sets in a statistical way and take some decisions in order to transform the data if is required (feature engineering).

In [1]:
# We use the default data manipulation library in python: pandas
import pandas as pd 
# Numpy as the standard computation tools over arrays and matrices 
import numpy as np

bank_df = pd.read_excel("data/banknote_data/data_banknote_authentication.xlsx")

occupancy_df = pd.read_excel("data/occupancy_data/occupancy_data.xlsx")

## Banknote authentication data set processing

In [2]:
# Fast first idea of the content of the data set
bank_df.head(2)

variance  skewness  curtosis  entropy  output
0    3.6216    8.6661   -2.8073 -0.44699       0
1    4.5459    8.1674   -2.4586 -1.46210       0

In [3]:
# First we give us an idea of the size of the dataset
bank_df.shape
# 1372 examples, 4 features and 1 output column

(1372, 5)

This tell us the data has 1372 examples, 4 feature columns and 1 label column. This is what we expect from our summary of the data. <br>

Remember, the labels in our data set are confusing: 0 for genuine banknotes and 1 for no genuine banknotes. This confuse our analysis of the performance metric about false positive and negative cases. For this reason, we need to swap the labels of our samples (1 label is swaped to 0 label, and vice versa).

In [4]:
# Swaping the 0 labels for 1 label, and vice versa
bank_df["output"] = np.where(bank_df["output"] == 1, 0, 1)

In order to comprobe if our performance metric is appropriate now we'll see if the data set is imbalanced:

In [5]:
# Group the df for each label and count the number of samples
temp_series = bank_df.groupby('output')['output'].count()
temp_df = pd.DataFrame({'classes':temp_series.index, 'count':temp_series.values})
temp_df['proportion'] = (temp_df['count']/bank_df.shape[0]).apply('{:.0%}'.format)
temp_df
# The result df give us the respective proportion of each label over the total of samples

classes  count proportion
0        0    610        44%
1        1    762        56%

We can see the proportion of each class in our data set is good. Moreover, for each sample labeled as 0, we have 1.2 samples labeled as 1. This is a balanced proportion of classes. Also, this means our data set has more observations of genuine banknotes than not genuine banknotes.<br>

Because the nature ot the data (each feature comes from a wavelet decomposition) we'll see the features has good behavior. <br>

Now, before we split the data in train-test form, we shuffle the data set to mostly guarantee an independent sample view to the algorithm.

In [6]:
# Shuffle the dataframe
bank_df = bank_df.sample(frac=1)

Split the data frame in a train-test form (70% train - 30% test)

In [7]:
# How much examples must be in the 70% training data
1372*0.7

960.4

This tell us we need take the first 961 rows to train, and the remaining rows for test.

In [8]:
# Split in train and test
bank_train = bank_df[:961]
bank_test = bank_df[961:]

Set up the data in the scikit-learn form.

In [9]:
# Create a matrix X_train with the features of shape (n_samples, n_features)
# Do the same for the real label of each sample in Y_train of shape (n_samples,)
# Repeat the procces for the test set

X_train_bank, Y_train_bank = bank_train.iloc[:,:4].values, bank_train.iloc[:,-1].values
X_test_bank, Y_test_bank = bank_test.iloc[:,:4].values, bank_test.iloc[:,-1].values

In [10]:
# Y_train actually have his respective number of examples? Yes
Y_train_bank.shape

(961,)

## Occupancy detection data set procesing

In [11]:
# Again, first we take a quick look of the data 
occupancy_df.head(2)

date  Temperature  Humidity  Light     CO2  HumidityRatio  \
0 2015-02-04 17:51:00        23.18   27.2720  426.0  721.25       0.004793   
1 2015-02-04 17:51:59        23.15   27.2675  429.5  714.00       0.004783   

   Occupancy  
0          1  
1          1

In [12]:
# See how many examples and features we have in our data set
occupancy_df.shape
# 20560 examples, 6 features and 1 output column

(20560, 7)

In [13]:
# Study the proportion of classes over the dataset
# Group the df for each label and count the number of samples
temp_series = occupancy_df.groupby('Occupancy')['Occupancy'].count()
temp_df = pd.DataFrame({'classes':temp_series.index, 'count':temp_series.values})
temp_df['proportion'] = (temp_df['count']/occupancy_df.shape[0]).apply('{:.0%}'.format)
temp_df
# The result df give us the respective proportion of each label over the total of samples

classes  count proportion
0        0  15810        77%
1        1   4750        23%

We can conclude our data set is mild imbalanced. Although, we'll keep forward to see if models generalize well with the original distribution of the data. <br> 

Now, we need to analyce each column to see if the features have good behavior or we need to manipulate them in some way. <br> 

First of all, note that the date column is in a non-processable form for the model. Let's study it a little bit to know what kind of date time information store.

In [14]:
# Year of data
occupancy_df.date.dt.year.unique()
# Only 2015 data

array([2015], dtype=int64)

In [15]:
# Month of data
occupancy_df.date.dt.month.unique()
# Only february data

array([2], dtype=int64)

In [16]:
# Days of data
occupancy_df.date.dt.day.unique()
# Some days of february

array([ 4,  5,  6,  7,  8,  9, 10,  2,  3, 11, 12, 13, 14, 15, 16, 17, 18],
      dtype=int64)

In [17]:
# Hours of data
occupancy_df.date.dt.hour.unique()
# We have records of all hours from the above days

array([17, 18, 19, 20, 21, 22, 23,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
       10, 11, 12, 13, 14, 15, 16], dtype=int64)

In [18]:
# Minutes of data
occupancy_df.date.dt.minute.unique()
# We have records of all minutes for each hour

array([51, 53, 54, 55, 57, 58,  0,  1,  2,  3,  4,  6,  7,  8, 10, 11, 13,
       14, 15, 16, 17, 19, 20, 21, 22, 23, 24, 26, 27, 28, 29, 30, 32, 33,
       34, 35, 36, 38, 39, 40, 42, 43, 45, 46, 47, 48, 49, 52, 56, 59,  5,
        9, 12, 18, 25, 31, 37, 41, 44, 50], dtype=int64)

Because we have only data from february 2015, this two date time elements are negligible to our model. Though days vary from 4 up to 18 of february, a more rich information is hours and minutes. Intuitively, the occupancy of a room has a correlation with the hour of the day; must be no the same for day or night. Remember the data was collected of records taken every minute, so, take this feature as the hour and minute of the day is a good handcrafted method. <br>
 
However, we are assuming the description we have of data is true, but to be sure, lets see if the number of examples for every hour is balanced.

In [19]:
# Create a copy of the occupancy dataframe to extract only the hour of the day from the date column
test_df = occupancy_df.copy()
test_df['hour'] = test_df.date.dt.hour

# Calculate the proportion of examples for each hour
temp_series = test_df.groupby('hour')['hour'].count()
temp_df = pd.DataFrame({'hour':temp_series.index, 'count':temp_series.values})
temp_df['proportion'] = (temp_df['count']/occupancy_df.shape[0]).apply('{:.0%}'.format)
temp_df

hour  count proportion
0      0    900         4%
1      1    915         4%
2      2    885         4%
3      3    900         4%
4      4    915         4%
5      5    885         4%
6      6    900         4%
7      7    915         4%
8      8    885         4%
9      9    834         4%
10    10    776         4%
11    11    708         3%
12    12    720         4%
13    13    732         4%
14    14    761         4%
15    15    840         4%
16    16    854         4%
17    17    835         4%
18    18    900         4%
19    19    915         4%
20    20    885         4%
21    21    900         4%
22    22    915         4%
23    23    885         4%

We can conclude the number of examples for each hour is balanced, moreover, the description we have over the data is correct. <br>

We decide to keep only with hours and minutes. So, the one date feature will become two date features: hour and minute.

In [20]:
# Create columns with hour and minute
occupancy_df['hour'] = occupancy_df['date'].dt.hour
occupancy_df['minute'] = occupancy_df['date'].dt.minute
# Drop the date column
occupancy_df.drop(columns='date', inplace = True)
# Re-arange the columns to keep the labels in last column
occupancy_df = occupancy_df[['Temperature', 'Humidity', 'Light', 'CO2', 'HumidityRatio', 'hour', 'minute', 'Occupancy']]

Since this point, the data process will be the same ast he banknote data set process. We're gonna shuffle tha data and extract train and test sets in 70% - 30% proportion.

In [21]:
# Shuffle the data
occupancy_df = occupancy_df.sample(frac=1)

In [22]:
# How much examples must be in the train set
20560*0.7
# Must be 1392 samples

14391.999999999998

In [23]:
# Split in train and test
occupancy_train = occupancy_df[:14392]
occupancy_test = occupancy_df[14392:]

In [24]:
# Create a matrix X_train with the features of shape (n_samples, n_features)
# Do the same for the real label of each sample in Y_train of shape (n_samples,)
# Repeat the procces for the test set

X_train_occupancy, Y_train_occupancy = occupancy_train.iloc[:,:7].values, occupancy_train.iloc[:,-1].values
X_test_occupancy, Y_test_occupancy = occupancy_test.iloc[:,:7].values, occupancy_test.iloc[:,-1].values

In [25]:
# Y_train actually have his respective number of examples? Yes
Y_train_occupancy.shape

(14392,)

# Models Implementation

## Logistic Regression

In [26]:
def LogisticRegression_train(X_train, Y_train):
    '''
    Import and fit a logistic regression model from scikit-learn.
    
    Input:
    X_train: Training X set of samples and features. Shape: (# of samples, # of features)
    Y_train: Training Y set of labels for the samples in X_train. Shape: (# of samples, )
    
    Output: Trained logistic regression model
    '''
    
    from sklearn.linear_model import LogisticRegression
    model = LogisticRegression().fit(X_train, Y_train)
    return model

In [27]:
# Fit a logistic model for both data sets 

# Banknotes model 
LogReg_bank = LogisticRegression_train(X_train_bank, Y_train_bank)

# Occupancy model
LogReg_occupancy = LogisticRegression_train(X_train_occupancy, Y_train_occupancy)

In [28]:
# Extract the predicted labels test and train sets for both data sets, banknotes and occupancy

# Banknotes set predicts 
LogReg_train_bank_predicts = LogReg_bank.predict(X_train_bank)
LogReg_test_bank_predicts = LogReg_bank.predict(X_test_bank)

# Occupancy set predicts
LogReg_train_occupancy_predicts = LogReg_occupancy.predict(X_train_occupancy)
LogReg_test_occupancy_predicts = LogReg_occupancy.predict(X_test_occupancy)

### Performance metrics by confusion matrix example

Rememeber, here we want to measure the performance of our model based in the precision score (PS) and for standar studies the accuracy score (AS). The following code cells calculate the PS and AS by seen the confusion matrix for our logistic model implemented above and manually calculate the PS and AS.

In [29]:
# Import the confusion matrix for sklearn and see it for the training set
from sklearn.metrics import confusion_matrix
confusion_matrix(Y_train_bank, LogReg_train_bank_predicts)

array([[429,   3],
       [  7, 522]], dtype=int64)

In [30]:
# We can also extract the TP, FP, FN, TP number of cases from the matrix 
tn, fp, fn, tp = confusion_matrix(Y_train_bank, LogReg_train_bank_predicts).ravel()
tn, fp, fn, tp 

(429, 3, 7, 522)

In [31]:
# Calculate the PS and AS for training set
PS = tp / (tp + fp)
AS = (tp + tn) / (tp + fp + fn + tn)
print('For banknote authentication training set, the AS is', AS, 'and the PS is', PS)

For banknote authentication training set, the AS is 0.9895941727367326 and the PS is 0.9942857142857143


In [32]:
# See the confusion matrix for test set
confusion_matrix(Y_test_bank, LogReg_test_bank_predicts)

array([[177,   1],
       [  3, 230]], dtype=int64)

In [33]:
# Extact the tp, fp, fn, tp cases and calculate the PS score for test set
tn, fp, fn, tp = confusion_matrix(Y_test_bank, LogReg_test_bank_predicts).ravel()
PS = tp / (tp + fp)
AS = (tp + tn) / (tp + fp + fn + tn)
print('For banknote authentication test set, the AS is', AS, 'and the PS is', PS)

For banknote authentication test set, the AS is 0.9902676399026764 and the PS is 0.9956709956709957


Now we'll see how to do the same process directly without calculate the confusion matrix. <br>

### Performance metrics in banknote authentication

In [34]:
# Import the PS and AS metrics from sklearn and 
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
# Compute the metrics for train set
PS = precision_score(Y_train_bank, LogReg_train_bank_predicts, average='binary')
AS = accuracy_score(Y_train_bank, LogReg_train_bank_predicts)
print('For banknote authentication training set, the AS is', AS, 'and the PS is', PS)

For banknote authentication training set, the AS is 0.9895941727367326 and the PS is 0.9942857142857143


We see both PS and AS are close to one in our training set. This means we have **low bias** in our model.

In [35]:
# AS and PS for test set
PS = precision_score(Y_test_bank, LogReg_test_bank_predicts, average='binary')
AS = accuracy_score(Y_test_bank, LogReg_test_bank_predicts)
print('For banknote authentication test set, the AS is', AS, 'and the PS is', PS)

For banknote authentication test set, the AS is 0.9902676399026764 and the PS is 0.9956709956709957


Both PS and AS in our test set are also close to one. This means we have also **low variance** because training and test errors are close. Moreover, note both metrics are better in our test set than in our training set, this along with low bias and low variance is a good insight of generalization.

### Performance metrics in occupancy problem 

In [36]:
# From sklearn import RS metric and calculate RS and AS for training set
from sklearn.metrics import recall_score
RS = recall_score(Y_train_occupancy, LogReg_train_occupancy_predicts)
AS = accuracy_score(Y_train_occupancy, LogReg_train_occupancy_predicts)
print('For occupancy detection training set, the AS is', AS, 'and the PS is', RS)

For occupancy detection training set, the AS is 0.9895775430794886 and the PS is 0.9954941423851006


We see **low bias** for both metrics in occupancy detection training set.

In [37]:
# Calculate RS and AS over the test set
RS = recall_score(Y_test_occupancy, LogReg_test_occupancy_predicts)
AS = accuracy_score(Y_test_occupancy, LogReg_test_occupancy_predicts)
print('For occupancy detection test set, the AS is', AS, 'and the PS is', RS)

For occupancy detection test set, the AS is 0.9871919584954605 and the PS is 0.9922589725545391


Also, we see **low variance** when we compare training and test performance metrics (are close for both AS and RS). Again, this can allow us to generalize the models to new data.

## Decision tree learning

In [38]:
def DecisionTree_train(X_train, Y_train):
    '''
    Import and fit decision tree learning model from scikit-learn.
    
    Input:
    X_train: Training X set of samples and features. Shape: (# of samples, # of features)
    Y_train: Training Y set of labels for the samples in X_train. Shape: (# of samples, )
    
    Output: Trained decision tree
    '''
    
    from sklearn.tree import DecisionTreeClassifier
    model = DecisionTreeClassifier().fit(X_train, Y_train)
    return model

In [39]:
# Fit a decision tree model for both data sets 

# Banknotes model 
DecTree_bank = DecisionTree_train(X_train_bank, Y_train_bank)

# Occupancy model
DecTree_occupancy = DecisionTree_train(X_train_occupancy, Y_train_occupancy)

In [40]:
# Extract the predicted labels test and train sets for both data sets, banknotes and occupancy

# Banknotes set predicts 
DecTree_train_bank_predicts = DecTree_bank.predict(X_train_bank)
DecTree_test_bank_predicts = DecTree_bank.predict(X_test_bank)

# Occupancy set predicts
DecTree_train_occupancy_predicts = DecTree_occupancy.predict(X_train_occupancy)
DecTree_test_occupancy_predicts = DecTree_occupancy.predict(X_test_occupancy)

### Performance metrics in banknote authentication

In [41]:
# Compute the metrics for train set
PS = precision_score(Y_train_bank, DecTree_train_bank_predicts, average='binary')
AS = accuracy_score(Y_train_bank, DecTree_train_bank_predicts)
print('For banknote authentication training set, the AS is', AS, 'and the PS is', PS)

For banknote authentication training set, the AS is 1.0 and the PS is 1.0


Note both performance metrics are close to one. This is, we have **low bias**. We can think perfect behavior (both metrics are one) wouldn't allow model generalize over new data.

In [42]:
# AS and PS for test set
PS = precision_score(Y_test_bank, DecTree_test_bank_predicts, average='binary')
AS = accuracy_score(Y_test_bank, DecTree_test_bank_predicts)
print('For banknote authentication test set, the AS is', AS, 'and the PS is', PS)

For banknote authentication test set, the AS is 0.9805352798053528 and the PS is 0.9828326180257511


However, the test performance metrics are close to train metrics. Mereover, both are close to one. This means we have **low variance**. Both, low bias and variance allow us to say we can generalize the model over new banknote data.

### Performance metrics in occupancy detection

In [43]:
# Compute RS and AS for training set
RS = recall_score(Y_train_occupancy, DecTree_train_occupancy_predicts)
AS = accuracy_score(Y_train_occupancy, DecTree_train_occupancy_predicts)
print('For occupancy detection training set, the AS is', AS, 'and the PS is', RS)

For occupancy detection training set, the AS is 1.0 and the PS is 1.0


Note the **low bias** of the model.

In [45]:
# Calculate RS and AS over the test set
RS = recall_score(Y_test_occupancy, DecTree_test_occupancy_predicts)
AS = accuracy_score(Y_test_occupancy, DecTree_test_occupancy_predicts)
print('For occupancy detection test set, the AS is', AS, 'and the PS is', RS)

For occupancy detection test set, the AS is 0.9920557717250325 and the PS is 0.9845179451090781


Both metrics are close to one and even close to them respective metric in the training set, this is **low variance**. Again, we can suppose the model will generalize over new data.

## Support vector machine (SVM)

For this implementation of support vector machines we're gonna do a grid search for the $C$ hyperparameter of soft SVM. $C$ acts as a penalty for missclasified data points for the decision bundary, allowing us make some errors in the training phase.
More information of SVM hyperparameters in: [sklearn SVC](https://scikit-learn.org/stable/modules/svm.html#svm-classification).

### Grid Search for SVM

A grid search consist of train systematically multiple models with all combinations of hyperparameters (for this example only one parameter). At the end, we keep with the hypterparameters that give us the model with better performance. This function is also found in sklearn.

In [120]:
def SVM_GridSearch(hyperparams_dict, score, X_train, Y_train):
    '''
    Import and fit support vector machine models with grid search from scikit-learn.
    
    Input:
    hyperparams_dict: Dict of hyperparameters for the SVM algorithm; params like C
    score: Metric to compare performances of trained SVM models
    X_train: Training X set of samples and features. Shape: (# of samples, # of features)
    Y_train: Training Y set of labels for the samples in X_train. Shape: (# of samples, )
    
    Output: Grid search findings for SVMs models trained
    '''
    
    from sklearn.model_selection import GridSearchCV
    from sklearn.svm import LinearSVC
    
    grid_search = GridSearchCV(LinearSVC(), 
                               hyperparams_dict, 
                               scoring = score,
                               return_train_score=True, 
                               cv = 5,
                               verbose = 1,
                               refit = True)
    
    return grid_search.fit(X_train, Y_train)

In [116]:
# We declare the hyperparameters dictionary containing all the C-values to explore in grid search
hyperparams_dict = {'C':[0.0001, 0.001, 0.01, 0.1, 1, 10, 100]}

# Fit the SVM for banknote authentication data set (remember the metric is precision)
score = 'precision'
SVM_bank = SVM_GridSearch(hyperparams_dict, score, X_train_bank, Y_train_bank)

# Fit the SVM for room occupancy data set (remeber the metric is recall)
score = 'recall'
SVM_occupancy = SVM_GridSearch(hyperparams_dict, score, X_train_occupancy, Y_train_occupancy)

Fitting 5 folds for each of 7 candidates, totalling 35 fits


C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathema

Fitting 5 folds for each of 7 candidates, totalling 35 fits


C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathematics\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\edwar\anaconda3\envs\ML_mathema

In [121]:
# Extract the predicted labels test and train sets for both data sets, banknotes and occupancy

# Banknotes set predicts 
SVM_train_bank_predicts = SVM_bank.predict(X_train_bank)
SVM_test_bank_predicts = SVM_bank.predict(X_test_bank)

# Occupancy set predicts
SVM_train_occupancy_predicts = SVM_occupancy.predict(X_train_occupancy)
SVM_test_occupancy_predicts = SVM_occupancy.predict(X_test_occupancy)

### Performance metrics in banknote authentication

In [125]:
# Extract the best parameter values found in grid search
best_parameters = SVM_bank.best_params_
print('The best parameter value was:', best_parameters)

# Compute the metrics for train set
PS = precision_score(Y_train_bank, SVM_train_bank_predicts, average='binary')
AS = accuracy_score(Y_train_bank, SVM_train_bank_predicts)
print('For banknote authentication training set, the AS is', AS, 'and the PS is', PS)

The best parameter value was: {'C': 0.1}
For banknote authentication training set, the AS is 0.9895941727367326 and the PS is 0.9942857142857143


Note both performance metrics are close to one. This is, we have **low bias**.

In [126]:
# AS and PS for test set
PS = precision_score(Y_test_bank, SVM_test_bank_predicts, average='binary')
AS = accuracy_score(Y_test_bank, SVM_test_bank_predicts)
print('For banknote authentication test set, the AS is', AS, 'and the PS is', PS)

For banknote authentication test set, the AS is 0.9927007299270073 and the PS is 1.0


However, the test performance metrics are close to train metrics. Mereover, both are close to zero. This means we have **low variance**. Both, low bias and variance allow us to say we can generalize the model over new banknote data.

### Performance metrics in occupancy detection

In [129]:
# Extract the best parameter values found in grid search
best_parameters = SVM_occupancy.best_params_
print('The best parameter value was:', best_parameters)

# Compute RS and AS for training set
RS = recall_score(Y_train_occupancy, SVM_train_occupancy_predicts)
AS = accuracy_score(Y_train_occupancy, SVM_train_occupancy_predicts)
print('For occupancy detection training set, the AS is', AS, 'and the PS is', RS)

The best parameter value was: {'C': 0.0001}
For occupancy detection training set, the AS is 0.9891606448026682 and the PS is 0.9966957044157405


Note the **low bias** of the model.

In [130]:
# Calculate RS and AS over the test set
RS = recall_score(Y_test_occupancy, SVM_test_occupancy_predicts)
AS = accuracy_score(Y_test_occupancy, SVM_test_occupancy_predicts)
print('For occupancy detection test set, the AS is', AS, 'and the PS is', RS)

For occupancy detection test set, the AS is 0.9878404669260701 and the PS is 0.9964813511611541


Both metrics are close to one and even close to them respective metric in the training set, this is **low variance**. Again, we can suppose the model will generalize over new data.

## K - Nearest neighbors (KNN)

In [71]:
def KNN_train(X_train, Y_train, n_neighbors=None):
    '''
    Import and fit KNN learning model from scikit-learn.
    
    Input:
    X_train: Training X set of samples and features. Shape: (# of samples, # of features)
    Y_train: Training Y set of labels for the samples in X_train. Shape: (# of samples, )
    n_neighbors: Number of neighbors the KNN algorothim use to compute the data label (sklearn default: 5)
    
    Output: Trained KNN model
    '''
    
    from sklearn.neighbors import KNeighborsClassifier
    model = KNeighborsClassifier(n_neighbors).fit(X_train, Y_train)
    return model

In [72]:
# Fit a decision tree model for both data sets with a K parameter of 2

# Banknotes model 
KNN_bank = KNN_train(X_train_bank, Y_train_bank, n_neighbors = 2)

# Occupancy model
KNN_occupancy = KNN_train(X_train_occupancy, Y_train_occupancy, n_neighbors = 2)

In [73]:
# Extract the predicted labels test and train sets for both data sets, banknotes and occupancy

# Banknotes set predicts 
KNN_train_bank_predicts = KNN_bank.predict(X_train_bank)
KNN_test_bank_predicts = KNN_bank.predict(X_test_bank)

# Occupancy set predicts
KNN_train_occupancy_predicts = KNN_occupancy.predict(X_train_occupancy)
KNN_test_occupancy_predicts = KNN_occupancy.predict(X_test_occupancy)

### Performance metrics in banknote authentication

In [76]:
# Compute the metrics for train set
PS = precision_score(Y_train_bank, KNN_train_bank_predicts, average='binary')
AS = accuracy_score(Y_train_bank, KNN_train_bank_predicts)
print('For banknote authentication training set, the AS is', AS, 'and the PS is', PS)

For banknote authentication training set, the AS is 0.9989594172736732 and the PS is 1.0


In [77]:
# AS and PS for test set
PS = precision_score(Y_test_bank, KNN_test_bank_predicts, average='binary')
AS = accuracy_score(Y_test_bank, KNN_test_bank_predicts)
print('For banknote authentication test set, the AS is', AS, 'and the PS is', PS)

For banknote authentication test set, the AS is 1.0 and the PS is 1.0


We have prefect behavior in training phase, that we can think as overfitting, but when we see test phase assess we see a very high performance too. We have a **low bias** and **low variance**. Therefore we can generalize.

### Performance metrics in occupancy detection

In [78]:
# Compute RS and AS for training set
RS = recall_score(Y_train_occupancy, DecTree_train_occupancy_predicts)
AS = accuracy_score(Y_train_occupancy, DecTree_train_occupancy_predicts)
print('For occupancy detection training set, the AS is', AS, 'and the PS is', RS)

For occupancy detection training set, the AS is 1.0 and the PS is 1.0


In [79]:
# Calculate RS and AS over the test set
RS = recall_score(Y_test_occupancy, DecTree_test_occupancy_predicts)
AS = accuracy_score(Y_test_occupancy, DecTree_test_occupancy_predicts)
print('For occupancy detection test set, the AS is', AS, 'and the PS is', RS)

For occupancy detection test set, the AS is 0.9920557717250325 and the PS is 0.9845179451090781


Again, we have **low bias** and **low variance**. We can generalize.

# Results

The finding results for our models will be resumed in the following tables, each one for each dataset:

**Banknote authentication data set**:

| Algorithm | $C$ | Precision in Train | Precision in test | Accuracy in train | Accuracy in test |
| --- | --- | --- | --- | --- | --- |
| Logistic Regression |  | 0.99 | 0.99 | 0.98 | 0.98 |
| Decision Tree |  | 1 | 0.98 | 1 | 0.98 |
| Support Vector Machine | 0.1 | 0.99 | 1 | 0.98 | 0.99 |
| K-Nearest Neighbors |  | 1 | 1 | 0.99 | 1 |

**Occupancy detection data set**:

| Algorithm | $C$ | Precision in Train | Precision in test | Accuracy in train | Accuracy in test |
| --- | --- | --- | --- | --- | --- |
| Logistic Regression |  | 0.99 | 0.99 | 0.98 | 0.98 |
| Decision Tree |  | 1 | 0.98 | 1 | 0.99 |
| Support Vector Machine | 0.0001 | 0.99 | 0.99 | 0.98 | 0.98 |
| K-Nearest Neighbors |  | 1 | 0.98 | 1 | 0.99 |

# Conclusions, Discussions and Future Work

We have some general conclusions:
<ol>
  <li>All the four models can generalize to new data.</li>
  <li>All the four models have almost perfect performance (some of them a perfect performance) in both, training and test sets.</li>
  <li>These data sets are good for learning to implement models, not for practice feature engineering (the data already have a ready to pass form).</li>
</ol>

Note: The possible results in new data of the chosen model depend on how much the new data resembles the training and test data. For these two particular datasets, we have the data in a highly processed form. For banknote authentication, computing the wavelet transform of banknote pictures may not be as fast as we think. Similarly, extracting all the features of a room to detect occupancy in real time can involve expensive infrastructure costs.

Real-world scenarios involve highly complex datasets. Preprocessing and performing valuable feature engineering are probably the most important and time-consuming aspects in an ML project. Future work can aim to explore more complex open datasets and iterate over a data methodology, such as CRIPS DM, to develop valuable business insights.